[1단계 - 로또(자동)] 조조그린(조한석) 미션 제출합니다.

jojogreen91 opened the pull request:
안녕하세요 저는 **조조그린**이라고 합니다!
반갑습니다 리뷰어님 :)

미션을 수행 하면서 로또 미션의 요구사항과 더불어서
지난번 미션에서 배웠던 부분도 반영해보려고 개인적으로 노력해봤는데 어떨지 모르겠네요.
특히 `Controller` 부분은 더 좋게 만들수 있는 여지가 있는것 같아서 지금은 좀 아쉬운것 같아요.

그럼 떨리는 마음으로 리뷰 기다리고 있겠습니다...잘 부탁드립니다😀

summerlunaa committed:
docs: 기능목록 작성

summerlunaa committed:
feat: 구입금액이 자연수인지 검증

summerlunaa committed:
feat: 구입금액이 1000원 단위인지 검증

summerlunaa committed:
feat: LottoTicket 생성 기능

summerlunaa committed:
feat: 구입 금액만큼 로또 티켓을 생성하는 기능

summerlunaa committed:
feat: 로또 티켓을 구입하는 기능

summerlunaa committed:
docs: 로또에서 로또 티켓으로 수정

summerlunaa committed:
fix: static 추가

summerlunaa committed:
refactor: shuffleLottoNumbers 메서드 위치 변경

summerlunaa committed:
refactor: LottoMachine 삭제 및 getTicketCount 이동

summerlunaa committed:
refactor: MoneyReceiver 생성

summerlunaa committed:
feat: 로또 번호 유효성 검사 기능

summerlunaa committed:
feat: 당첨 번호 생성 유효성 검사

summerlunaa committed:
chore: 줄바꿈

summerlunaa committed:
fix: 로또 번호가 정렬되도록 수정

summerlunaa committed:
test: receiver에 대한 테스트 추가

summerlunaa committed:
feat: 보너스 번호 생성 유효성 검사

summerlunaa committed:
refactor: 문자열 상수화

summerlunaa committed:
chore: DisplayName 변경

summerlunaa committed:
feat: 당첨 번호와 일치하는 티켓의 번호 개수 구하는 기능

summerlunaa committed:
feat: 티켓에 보너스 번호가 존재하는지 확인하는 기능

summerlunaa committed:
refactor: 멤버변수 자료형 변경

summerlunaa committed:
refactor: 테스트 이름 변경

summerlunaa committed:
chore: 오타 수정

summerlunaa committed:
feat: 당첨 번호 매치 개수와 보너스 번호 매치 여부로 Rank 객체 찾기

summerlunaa committed:
fix: 당첨이 안되는 상황 추가

summerlunaa committed:
refactor: 로또 번호 자동 생성 로직을 LottoNumber 로 이동

summerlunaa committed:
feat: 당첨 번호와 보너스 번호를 이용해서 로또 티켓의 등수를 반환하는 기능

summerlunaa committed:
feat: 로또 티켓들의 당첨 갯수를 구하는 기능

jojogreen91 committed:
refactor: ticket에서 lotto로 이름 변경

jojogreen91 committed:
feat: 로또 자동, 수동 생성 정적 팩토리 매서드 추가

jojogreen91 committed:
refactor: 당첨 번호와 보너스 번호를 WinningNumbers 객체로 통합

jojogreen91 committed:
refactor: winningNumberReceiver가 LottoReceiver로 변경

jojogreen91 committed:
refactor: ticket에서 lotto로 이름 변경

jojogreen91 committed:
refactor: public에서 private로 변경

jojogreen91 committed:
test: DisplayName 변경

jojogreen91 committed:
refactor: 메서드 이름 변경

jojogreen91 committed:
refactor: 라인 순서 변경

jojogreen91 committed:
refactor: change names

jojogreen91 committed:
feat: 수익률을 계산하는 기능

jojogreen91 committed:
feat: 사용자에게서 구입금액을 입력받는 기능

jojogreen91 committed:
feat: 구매한 로또들을 출력하는 기능

jojogreen91 committed:
feat: 당첨 번호와 보너스 번호를 입력받는 기능

jojogreen91 committed:
refactor: 클래스 이름 변경

jojogreen91 committed:
refactor: 변수 이름 변경

jojogreen91 committed:
refactor: 인라인화

jojogreen91 committed:
feat: 당첨통계를 출력하는 기능

summerlunaa committed:
feat: 당첨 개수 출력 기능

summerlunaa committed:
fix: 오타 수정

summerlunaa committed:
feat: 총 수익률 출력 기능

summerlunaa committed:
refactor: 메서드 이름 변경

summerlunaa committed:
refactor: 출력 메서드 통합

summerlunaa committed:
refactor: 메서드 이름 변경

summerlunaa committed:
refactor: 스트림 사용

jojogreen91 committed:
refactor: Receiver 클래스들에서 Validator 클래스를 분리

jojogreen91 committed:
refactor: WinningNumbers 를 WinningLotto 와 BonusNumber 두 클래스로 분리

jojogreen91 committed:
test: 바뀐 메서드에 대응하여 테스트 수정

jojogreen91 committed:
test: 테스트 수정

jojogreen91 committed:
refactor: 메서드 구현을 스트림으로 변경

jojogreen91 committed:
refactor: 메서드 이름 변경

jojogreen91 committed:
refactor: 메서드 이름 변경

jojogreen91 committed:
docs: 기능 목록 업데이트

ddaaac commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,41 @@
+    private final List<LottoNumber> lottoNumbers;
+
+    private Lotto(List<LottoNumber> lottoNumbers) {
+        this.lottoNumbers = Collections.unmodifiableList(lottoNumbers);
```
방어적 복사를 활용해주셨어요 👍

두 가지 질문을 드릴 수 있을 거 같아요.
- 만약 생성자에서 넘어온 `lottoNumbers`를 외부에서 수정하면 어떤 일이 생길까요?
- `unmodifiableList()`는 불변일까요?

[방어적 복사와 불변](https://tecoble.techcourse.co.kr/post/2021-04-26-defensive-copy-vs-unmodifiable/)에 대한 글을 읽어보세요.

jojogreen91 replied:
`만약 생성자에서 넘어온 lottoNumbers를 외부에서 수정하면 어떤 일이 생길까요?`
-> unmodifiableList 는 새로운 객체를 생성하는 것이 아니기 때문에 원본에서 수정이 발생하면 같이 변경됩니다!
`unmodifiableList()는 불변일까요?`
-> unmodifiableList 는 불변을 보장 하지 않습니다!

---

범블비가 알려주신 링크를 참고해서 코드를 수정해봤는데요.

`Lotto` 생성자에서 멤버변수를 초기화 할 때
`new ArrayList<>(lottoNumbers)` 를 사용하도록 수정했습니다! (새로운 객체를 생성하는 방어적 복사)

그리고 `getLottoNumbers()` 로 인해 멤버변수 `lottoNumbers` 가 변경되는 것을 막기 위해
`unmodifiableList` 를 사용해서 반환해주도록 수정했습니다!

*제가 설정한 `LottoNumber` 가 `enum` 으로 생성됐고 특성도 사실상 불변객체같아서 이런 방식들을 더 효과적으로 사용할 수 있다고 생각합니다.

ddaaac commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,57 @@
+
+public enum LottoNumber {
+
+    NUMBER_1(1), NUMBER_2(2), NUMBER_3(3), NUMBER_4(4), NUMBER_5(5),
+    NUMBER_6(6), NUMBER_7(7), NUMBER_8(8), NUMBER_9(9), NUMBER_10(10),
+    NUMBER_11(11), NUMBER_12(12), NUMBER_13(13), NUMBER_14(14), NUMBER_15(15),
+    NUMBER_16(16), NUMBER_17(17), NUMBER_18(18), NUMBER_19(19), NUMBER_20(20),
+    NUMBER_21(21), NUMBER_22(22), NUMBER_23(23), NUMBER_24(24), NUMBER_25(25),
+    NUMBER_26(26), NUMBER_27(27), NUMBER_28(28), NUMBER_29(29), NUMBER_30(30),
+    NUMBER_31(31), NUMBER_32(32), NUMBER_33(33), NUMBER_34(34), NUMBER_35(35),
+    NUMBER_36(36), NUMBER_37(37), NUMBER_38(38), NUMBER_39(39), NUMBER_40(40),
+    NUMBER_41(41), NUMBER_42(42), NUMBER_43(43), NUMBER_44(44), NUMBER_45(45);
+
```
1~45까지 모든 값을 만들어주셨군요! 😮👍

`Integer` 객체 안에서 값을 어떻게 캐싱하고 있는지 한 번 보시고 참고할 수도 있을 거 같아요.

jojogreen91 replied:
`LottoNumber` `enum` 클래스에 45개의 고유한 `LottoNumber` 를 상수로서 생성했고
`findByNumber(int number)`(매개변수에 대한 유효성 검사도 포함되어 있음) 를 통해 `LottoNumber` 의 인스턴스를 반환하는 형태이기 때문에 캐싱한 것과 같은 효과라고 생각했는데 혹시 아닐까요?

ddaaac replied:
잘 구현해주셨어요. 위 코멘트는 정말 참고해보라고 남겼습니다.
하드코딩된 코드는 확장성이 떨어지고, 실수를 할 여지가 있으니 `Integer` 클래스의 캐시를 보고 저런 방법도 있구나 정도로 학습하시면 될 거 같아요. (로또에서는 확장의 여지가 없으니 수정할 필요는 없어보입니다)

jojogreen91 replied:
넵! 덕분에 캐싱에 대해서 다시 공부해 볼 수 있었습니다 감사합니다 범블비 :)

ddaaac commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,57 @@
+
+    public static List<LottoNumber> getRandomLottoNumbers() {
+        return IntStream.range(0, PICKUP_COUNT)
+                .mapToObj(LottoNumber.shuffleLottoNumbers()::get)
```
이쯤 되면 레퍼런스보다는 일반적인 람다 형태로 작성하는게 코드 가독성에 도움이 되지 않을까 싶어요 :-) 

추가로, stream().limit()을 사용해보는 것도 좋을 것 같네요.

jojogreen91 replied:
수정했습니다🔧

ddaaac commented on src/main/java/lotto/exception/MoneyException.java:
```diff
@@ -0,0 +1,11 @@
+package lotto.exception;
+
+public class MoneyException extends IllegalArgumentException{
```
포매팅을 한 번 해야될 거 같아요.

jojogreen91 replied:
어떤 포매팅을 말씀하시는지 잘 모르겠습니다...ㅠㅠ
혹시 조금만 더 힌트를 주실 수 있을까요?😂

ddaaac replied:
인텔리제이의 [reformat](https://www.jetbrains.com/help/idea/reformat-and-rearrange-code.html) 기능을 사용하라는 코멘트였는데 제가 잘못 전달했네요. 죄송합니다 😅

ddaaac replied:
여기 코멘트 한 번 확인해주세요~

ddaaac commented on src/main/java/lotto/Application.java:
```diff
@@ -0,0 +1,11 @@
+public class Application {
+
+    public static void main(String[] args) {
+        new Controller().run(new Scanner(System.in));
```
스캐너를 여기서 생성하는 이유가 있을까요?
InputView에서 갖고 있어도 되지 않을까 싶어서요.

jojogreen91 replied:
수정했습니다🔧

ddaaac commented on src/test/java/lotto/domain/RankTest.java:
```diff
@@ -0,0 +1,25 @@
+        Assertions.assertThat(Rank.getRank(winningNumbersMatchCount, bonusNumberMatch))
+                .isEqualTo(Rank.RANK_OUT);
+    }
+}
```
테스트 케이스가 더 있으면 좋을 거 같아요. 예를 들면 `4,true` 같은 경우가 있겠네요.

jojogreen91 replied:
테스트 케이스 추가했고 확인해 보니 `getRank()` 구현도 잘못된 부분이 있어서 수정했습니다!😂

ddaaac commented on src/main/java/lotto/receiver/BonusNumberReceiver.java:
```diff
@@ -0,0 +1,16 @@
+import lotto.receiver.validator.BonusNumberValidator;
+import lotto.receiver.validator.LottoNumberValidator;
+
+public class BonusNumberReceiver {
```
Receiver, Validator, 도메인 객체가 모두 분리된 이유에 대해서 들어보고 싶어요!

jojogreen91 replied:
receiver 와 validator 를 분리시킨 이유는
제가 만든 `validator` 중에 `LottoNumberValidator` 부분이 여러군데에서 겹친다고 생각했기 때문입니다.
'1 ~ 45 사이 정수' 라는 로또 번호에 대한 검증은 `LottoNumber` `enum` 클래스에서 `LottoNumber` 를 찾는데에도 쓰이고
보너스 번호를 입력받을 때, 당첨 로또 번호를 입력받을 때에도 필요한 검증 이기 때문에 `LottoNumberValidator` 를 분리했고
다른 `validator` 들도 통일성과 간결성을 위해서, 또 다른 부분들에서 검증 기능만 가져와 사용할 가능성을 고려하여 분리했습니다!

`receiver`, `validator` 부분을 도메인 영역과 분리한 이유는
입력을 받아 유효성을 검증해주고 도메인 형태 혹은 사용하기 좋은 형태로 변환을 해주는 기능은 도메인의 영역보다 단순한 Util적인 부분이라고 생각해서 분리했습니다!

ddaaac replied:
<img width="631" alt="image" src="https://user-images.githubusercontent.com/43440195/155836229-2dce6df4-bc35-48f8-abe6-984b2931a44c.png">
<img width="696" alt="image" src="https://user-images.githubusercontent.com/43440195/155836294-522595b8-da81-46ca-974b-b323850801f5.png">

이 두 부분을 말씀해주신 거 같아요.
말씀해주신대로 `LottoNumber#findByNumber`에서 이미 검증을 하고 있는데 다른 객체에서 `LottoNumber`를 사용할 때 한번 더 검증을 할 필요가 있을까요?

객체를 사용하는 이유 중 하나는 도메인 로직(검증 등등)이 해당 객체 내부에 자리잡아서 코드 중복을 줄일 수 있다는 점입니다. 이를 활용하면 좀 더 전체적인 구조를 단순하게 가져갈 수 있을 거 같아요.

ddaaac replied:
다시 보니 String -> 도메인 변환을 Receiver에서 하고 있군요. 팩토리 메서드 격으로 이런 클래스가 있는 건 좋은 거 같아요.

위 코멘트를 조금 더 보충해볼게요.

검증로직이 객체 내부에 있는 것과 객체 외부에 있는 경우를 나눠서 생각해보면 좋을 거 같아요.
Money 클래스를 예로 들면
```java
public class Money {

    private final int money;

    public Money(int value) {
        validate(value);
        this.money = value;
    }

    public Money(String input) {
        this(toInt(input));
    }
}
```
와 같이 객체 내부에서 검증을 하는 경우에 이 생성자([주 생성자](https://jackjeong.tistory.com/148)라고 합니다)를 사용하게 되면 언제나 동일하게 객체를 검증할 수 있습니다. 이 생성자를 사용하는 쪽에서는 내부에서 검증이 일어나는지조차 알 필요가 없겠죠.

반면에 객체 외부(정적 팩토리 메서드나 Receiver 모두 여기에 포함될 거 같아요)에 검증 로직이 있으면 외부 객체는 객체를 생성하기 전에 검증이 필요하단 사실을 알아야합니다. 어떤 검증 로직이 실행돼야하는지도 알아야하고요. 

위의 이유로 전 주생성자에서 검증을 수행하는 걸 선호하고 부생성자나 팩토리 메서드 등은 주 생성자를 활용하는 방식으로 구현하는 편이에요.

jojogreen91 replied:
> 말씀해주신대로 LottoNumber#findByNumber에서 이미 검증을 하고 있는데 다른 객체에서 LottoNumber를 사용할 때 한번 더 검증을 할 필요가 있을까요?

그렇네요! 지금 중복적으로 검증을 하고 있네요...
`findByNumber()` 으로 열거체를 불러올 때 검증하는것으로 정하고 중복되는 부분들은 제거 했습니다!

---

범블비의 피드백을 보고 `validator` 와 `receiver` 부분을 리팩토링 해봤는데요.

범블비가 말씀해주신 주생성자 라는 개념을 생각해보니 확실히 주생성자에서 객체의 생성-유효성 검사를 하고
정적 팩토리 메소드나 부생성자에서는 주생성자를 통해 객체를 생성한다면 더 간결한 구조가 되는거 같아요!

그래서 `validator` 를 주생성자에서 사용하기 좋도록 변경했고 `receiver` 에서 하던 형태를 변환하던 기능은
그냥 해당 도메인 안에 팩토리 메소드가 가져도 될 기능인것 같아서 `receiver` 는 삭제했습니다.

콘솔로 입력된 값으로 객체를 생성할 때 객체의 생성-유효성을 어떤 위치에서 해줘야 할지 항상 고민 됐었는데
주생성자라는 개념을 적용하니 뭔가 제가 찾던 답이 여기 있었던것 같네요.

범블비가 말씀해주신게 혹시 이런 방향이 맞나요? 🥲

ddaaac commented on src/main/java/lotto/receiver/validator/LottoNumberValidator.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    private static boolean isNaturalNumber(String input) {
+        return Pattern.compile(REGEX_FOR_NATURAL_NUMBER).matcher(input).find();
```
직접 정규표현식을 만드는 방법도 좋지만 충분히 안정된 API인 Integer.parseInt() + 범위 검증 쪽이 좀 더 가독성이 좋지 않을까 싶어요.
정규표현식을 사용하면 항상 한 눈에 들어오지 않는 부분과 테스트가 고민되더라구요.

jojogreen91 replied:
수정했습니다🔧

ddaaac commented on src/main/java/lotto/domain/RankCount.java:
```diff
@@ -0,0 +1,38 @@
+
+    public RankCount(Lottos lottos, WinningLotto winningLotto, BonusNumber bonusNumber) {
+        Map<Rank, Integer> rankCount = new HashMap<>();
+        Arrays.stream(Rank.values())
```
인스턴스 변수는 인터페이스를 사용하고 구현체를 넣어주는 방식 좋네요 👍

추가로 [EnumMap](https://www.baeldung.com/java-enum-map) 에 대하여 알아볼까요?

jojogreen91 replied:
오..EnumMap 이라는 것도 있군요! 😀
수정했습니다🔧

ddaaac commented on src/main/java/lotto/domain/RankCount.java:
```diff
@@ -0,0 +1,38 @@
+
+    private void increaseCount(Rank rank) {
+        rankCount.put(rank, rankCount.get(rank) + 1);
+    }
```
rankCount.compute() 를 사용해 볼 수 있을 거 같아요.

jojogreen91 replied:
수정했습니다🔧

ddaaac commented on src/main/java/lotto/domain/RankCount.java:
```diff
@@ -0,0 +1,38 @@
+                .forEach(rank -> rankCount.put(rank, 0));
+        this.rankCount = rankCount;
+        calculateRankCount(lottos, winningLotto, bonusNumber);
+    }
```
rankCount에는 계산이 완료된 map만 마지막에 할당하면 어떨까요?
지금은 rankCount put 해주는 부분이 숨겨져있어서 로직을 놓칠 수도 있을 거 같아요.

jojogreen91 replied:
생성자에서 `Enum` 클래스 모두를 `put` 해주기 때문에 괜찮다고 생각했는데
혹시나 `put` 되지 않은 원소를 `calculateRankCount` 에서 사용할 가능성을 예방하자는 말씀 맞을까요? 😀
수정했습니다🔧

ddaaac replied:
```java
    public RankCount(Lottos lottos, WinningLotto winningLotto, BonusNumber bonusNumber) {
        this.rankCount = calculateRankCount(lottos, winningLotto, bonusNumber);
    }
```
위 코드처럼 rankCount 인스턴스 변수를 수정하지 않고 모두 계산이 끝난 뒤에 할당하면 좋겠다는 코멘트였습니다 😀
이렇게 하면 rankCount를 불변 Map으로 가져갈 수도 있구요.

jojogreen91 replied:
아 그렇네요!😂
수정했습니다🔧

ddaaac commented on src/main/java/lotto/domain/WinningLotto.java:
```diff
@@ -0,0 +1,20 @@
+        return (int) winningLotto.getLottoNumbers().stream()
+                .filter(lotto.getLottoNumbers()::contains)
+                .count();
+    }
```
getter를 사용하지 않고 구현할 수도 있지 않을까요?

jojogreen91 replied:
수정했습니다🔧
getter 를 최대한 안쓰려고 노력하는데 항상 어려운거 같아요...상황에 따라서는 어느쪽에 getter 를 만들어야 될지도 고민되고ㅠ
이번 미션을 하면서는 getter 를 쓰더라도 반환값으로 인해 원본 객체의 상태가 변하지 않도록 신경쓰긴 했는데 어떨지 모르겠습니다.

ddaaac replied:
잘 반영해주셨습니다 💯 

`WinningLotto`를 조합을 사용해서 잘 구현해주셔서 조금만 더 코드를 수정하면 `WinningLotto`의 getter도 제거할 수 있을 거 같아요.
지금은 `Lotto`에서 해당 getter를 사용하고 있는데 `WinningLotto`가 보너스 번호까지 포함한 객체가 되면 `Rank`를 구하는 로직도 `WinningLotto`로 옮겨질 수 있겠죠. 그러면 `WinningLotto`의 getter가 사라질 수 있을 거 같은데 시간이 되시면 한 번 구현해보시는 것도 추천드립니다 :)

jojogreen91 replied:
처음 구현할 때 저도 `getter` 나 불필요한 메서드를 줄이고 싶어서 `WinningLotto` 와 `BonusNumber` 가 합쳐진 형태의 클래스를 만들어서 구현 했었는데
문제가 `Controller` 에서 입력을 받을 때 `Exception` 이 발생하면 당첨 번호와 보너스 번호를 같이 다시 입력 받아야 하는 형태가 되더라구요.(다시 같이 초기화 해줘야 해서)
그래서 마지막에 다시 분리 했는데 그래도 둘이 합쳐서 `getter` 를 최대한 안쓰는 쪽으로 가는것이 더 좋을까요?

ddaaac replied:
꼭 보너스 번호가 잘못되었을 때 매번 다시 입력받을 필요가 있을까요?
당첨번호와 보너스 번호를 하나의 단위로 본다면 같이 입력받는게 좀 더 올바른 방향일 수도 있을거라는 생각이들어요.

ddaaac commented on src/main/java/lotto/controller/Controller.java:
```diff
@@ -0,0 +1,54 @@
+        RankCount rankCount = new RankCount(lottos, winningLotto, bonusNumber);
+        ProfitRate profitRate = new ProfitRate(rankCount.getTotalPrize(), money);
+        OutputView.printWinningStatistic(rankCount, profitRate);
+    }
```
컨트롤러 작성 잘 해주셨어요.

동작 단위로 적절하게 메서드 분리가 된다면 보기 더 편할 거 같지 않나요?
로또를 사고 -> 당첨 번호를 입력받고 -> 결과를 출력한다
이런 식으로 적절히 나누면 조금 더 읽기 좋은 코드가 될 것 같습니다 :)

jojogreen91 replied:
수정했습니다🔧

ddaaac commented on src/test/java/lotto/receiver/LottoReceiverTest.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @ParameterizedTest(name = "숫자가 6개가 아닌 경우 예외 발생 - case : {0}")
+    @ValueSource(strings = {"1, 2, 3, 4, 5", "1, 2, 3, 4, 5, 6, 7"})
```
엣지 케이스 검증 👍
테스트 코드들 전체적으로 잘 구현해주셨습니다!

jojogreen91 replied:
감사합니다😋

ddaaac commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -0,0 +1,15 @@
+
+public class LottoTest {
+
+    @Test
```
DisplayName이 빠져있는 테스트들이 있는데 한 번 확인 부탁드려요 :-)

jojogreen91 replied:
수정했습니다🔧

ddaaac commented on src/main/java/lotto/exception/LottoException.java:
```diff
@@ -0,0 +1,11 @@
+package lotto.exception;
+
+public class LottoException extends IllegalArgumentException {
```
커스텀 예외 잘 만들어주셨습니당

jojogreen91 replied:
감사합니다😋

ddaaac reviewed (changes requested):
안녕하세요 조조그린 👋

잘 구현해주셨습니다. 몇 가지 코멘트들과 질문사항 남겼으니 확인 부탁드려요!

궁금하신 점은 언제든 편하게 DM 주세요.

jojogreen91 committed:
fix: Lotto 생성자와 getLottoNumbers() 수정

jojogreen91 committed:
refactor: 메서드 로직 변경

jojogreen91 committed:
refactor: Scanner 생성 위치 변경

jojogreen91 committed:
fix: getRank() 로직 수정과 관련된 테스트 수정

jojogreen91 committed:
refactor: 로또 번호를 가지고 올때 Integer.parseInt 를 통해 정수인지 판별하도록 변경

jojogreen91 committed:
test: 누락된 부분 수정

jojogreen91 committed:
refactor: run() 메서드 리팩토링

jojogreen91 committed:
refactor: RankCount 리팩토링
- putIfAbsent, computeIfPresent, getOrDefault 활용

jojogreen91 committed:
refactor: RankCount 생성자 수정

jojogreen91 committed:
refactor: getLotto 삭제

jojogreen91 committed:
chore: 불필요한 파일 삭제

jojogreen91 committed:
docs: 기능목록 업데이트

jojogreen91 commented:
안녕하세요 범블비!
꼼꼼한 리뷰 감사드려요 :)
범블비가 남겨주신 피드백들을 반영해봤고 관련된 질문도 댓글로 남겼습니다.
다음 리뷰도 잘 부탁드립니다. 좋은 주말 보내세요 범블비😀

ddaaac commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -0,0 +1,26 @@
+        return scanner.nextLine();
+    }
+
+    public static String inputWinningLotto() {
```
개인적인 의견으로는 여기서 `List<Integer>`까지는 반환해줄 수 있지 않을까 싶어요.
`LottoNumber` 등은 도메인이니 View에서 분리한다고 하더라도, View에서 정수타입의 Integer를 넘겨줘야한다는 사실은 알아도 되지 않을까요?

jojogreen91 replied:
위쪽의 피드백을 반영하면서 현재 `WinningLotto` 에 `generateWinningLottoByConsole(String consoleInput)` 이라는 정적-팩토리-메서드를 만들었는데
이 정적-팩토리-메서드가 콘솔에서 넘어온 문자열로 `WinningLotto` 를 생성하는 기능이라고 생각하면 `inputWinningLotto()` 가 `String` 을 반환해도 괜찮다고 생각이 드는데 어떠신가요?😀

ddaaac replied:
개인적으로는 도메인 객체가 문자열을 파싱하는 책임까지 가져갈 필요는 없다고 생각하는 편입니다.
하지만 조조그린의 의도처럼 문자열 파싱의 책임까지 객체가 가져간다고 하면 지금 구조도 괜찮다고 생각해요 😀

ddaaac commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,56 @@
+        return Arrays.stream(LottoNumber.values()).collect(Collectors.toList());
+    }
+
+    public static LottoNumber findByNumber(int number) {
```
이 public 메서드가 외부에서 가장 많이 사용하게 될 메서드인 것 같은데 제일 위에 있는 건 어떨까요?

jojogreen91 replied:
그게 더 맞는거 같아요!
수정했습니다🔧

ddaaac commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,56 @@
+
+    public static LottoNumber findByNumber(int number) {
+        LottoNumberValidator.validate(number);
+        return LottoNumber.valueOf(LOTTO_NUMBER_PREFIX + number);
```
검증을 명시하기엔 validate를 거치는 것도 괜찮아보이는데 valueOf()에서 값이 없으면 던지는 `IllegalArgumentException`을 잡아서 커스텀 예외로 변환해주는 것만으로도 충분해보이기도 하네요.
조조그린은 어떤 쪽이 더 좋은 것 같나요?

jojogreen91 replied:
LottoNumberValidator 가 지금 체크해주고 있는 조건이 하나 뿐이라서 범블비가 말씀해주신 방법도 간결하고 좋은것 같네요!
수정했습니다🔧

ddaaac reviewed (changes requested):
안녕하세요 조조그린 👋

리뷰 반영 잘 해주셨어요! 코드가 깔끔해지고 있는 게 눈에 띄네요.

몇 가지 코멘트 남겼으니 확인해주시고 궁금한 점은 언제든지 DM 주세요 :)

jojogreen91 committed:
fix: 중복되는 검증 제거

jojogreen91 committed:
refactor: 도메인 객체 유효성 검사 위치 변경
- 주생성자를 만들어서 주생성자에서 객체의 유효성 검사를 하도록 변경
- 부생성자, 정적 팩토리 메소드는 주생성자를 통해 생성하도록 변경
- receiver 제거
- validator 를 주생성자에서 사용할 수 있도록 변경

jojogreen91 committed:
test: 변경된 프로덕션 코드에 맞춰 테스트코드 변경

jojogreen91 committed:
refactor: RankCount 생성자 수정

jojogreen91 committed:
refactor: RankCount 생성자 unmodifiableMap 추가

jojogreen91 committed:
refactor: RankCount 생성자 수정

jojogreen91 committed:
refactor: 생성자의 초기화 부분 수정

jojogreen91 committed:
refactor: getMatchCount 매개변수 변경

jojogreen91 committed:
refactor: LottoNumberValidator 삭제

jojogreen91 commented:
안녕하세요 범블비, 세번째 리뷰요청 보냅니다👋

범블비가 남겨주신 피드백들 반영해봤고 관련해서 댓글들도 남겨놨습니다!

이번 피드백을 반영하면서 뭔가 많이 배운것 같네요😀

이번에도 잘부탁드리고 감사합니다 범블비 :)

jojogreen91 committed:
refactor: printErrorMessage 로직 수정

ddaaac commented on src/main/java/lotto/domain/WinningLotto.java:
```diff
@@ -0,0 +1,22 @@
+        this.winningLotto = winningLotto;
+    }
+
+    public static WinningLotto generateWinningLottoByConsole(String consoleInput) {
+        return new WinningLotto(Lotto.generateLottoByConsole(consoleInput));
+    }
```
콘솔로 특정지은 이유가 있을까요?

ddaaac commented on src/main/java/lotto/domain/RankCount.java:
```diff
@@ -0,0 +1,39 @@
+    private final Map<Rank, Integer> rankCount;
+
+    public RankCount(Lottos lottos, WinningLotto winningLotto, BonusNumber bonusNumber) {
+        this.rankCount = new EnumMap<>(calculateRankCount(lottos, winningLotto, bonusNumber));
```
여기서도 방어적 복사가 필요할까요?

ddaaac reviewed (approved):
안녕하세요 조조그린 👋

리뷰 반영 잘 해주셨습니다. 이번 단계는 머지할게요.

몇 가지 코멘트 남겼으니 다음 단계에 반영해주세요 :)
